### ISSUE TITLE
Extension to 3D Wavelet Transform?

### ISSUE THREAD

**Issue by homerjed** [fbcotter/pytorch_wavelets#31]
Hi, Is there any future work planned toward allowing 3D DWT transforms? E.g. on an array of shape `(batch, channels, height, width, depth) `. 

I'd be very interested in working on this myself if you know of any good theoretical literature etc. Cheers!

**Comment by fbcotter** [comment]
Hey @homerjed I hadn't planned on extending to 3D DWTs, but in practice it wouldn't be too tricky to do! There's also a 3D DTCWT available too but that may be slightly harder to implement. 

As for any literature, I haven't seen much myself but I know that MRI analysis would often use 3d wavelet transforms, e.g. [[1]](https://asp-eurasipjournals.springeropen.com/articles/10.1186/1687-6180-2012-40), [[2]](https://tel.archives-ouvertes.fr/file/index/docid/772308/filename/minhphuong.tran_2569_vm.pdf) and [[3]](https://www-sigproc.eng.cam.ac.uk/foswiki/pub/Main/NGK/Chen_H_TIP_effreg_14may2011_submitted.pdf). These are all old school Signal processing techniques though, perhaps there's some room for deeplearning on top of these too?

**Comment by homerjed** [comment]
Thanks for your reply and the recommendations! The main thing I am looking to do is a Haar 3D-DWT as part of a parallelised layer in `Pytorch`. 

I think all I need is the Haar wavelets in 3D as a kind of matrix/vector form; i.e. each 'lll, llh, lhl...hhh' as I have seen it expressed before. These objects would be the registered buffers of the layer, like the parameters of a basic `batchnorm` layer. 

Sorry if I have confused any concepts here, I can't find anything on this exact issue, but have a working 2D layer that is equivalent.

**Comment by RY-97** [comment]
> Hi, Is there any future work planned toward allowing 3D DWT transforms? E.g. on an array of shape `(batch, channels, height, width, depth) `.
> 
> I'd be very interested in working on this myself if you know of any good theoretical literature etc. Cheers!



> Hi, Is there any future work planned toward allowing 3D DWT transforms? E.g. on an array of shape `(batch, channels, height, width, depth) `.
> 
> I'd be very interested in working on this myself if you know of any good theoretical literature etc. Cheers!

I was doing a similar job recently and encountered the same problem. I would like to ask for your help. I want to input a (b, c, w, h, d) 3D image data. The output after discrete wavelet transform is still consistent in shape, but does not require inverse wavelet transform. Do you have any solution?
